### PR TITLE
send job notification templates _after_ all events have been processed

### PR DESCRIPTION
**Pull request by ryanpetrello**
see: https://github.com/ansible/awx/issues/500

### CODE REVIEW

File: awx/main/management/commands/run_callback_receiver.py [hunk 1]
```diff
@@ -161,14 +162,35 @@ def _save_event_data():
...
+                                if uj.finished:
+                                    uj.send_notification_templates('succeeded' if uj.status == 'successful' else 'failed')
+                                    break
+                                else:
+                                    # wait a few seconds to avoid a race where the
+                                    # events are persisted _before_ the UJ.status
+                                    # changes from running -> successful
+                                    retries += 1
+                                    time.sleep(1)
+                                    uj = UnifiedJob.objects.get(pk=job_identifier)
```
Comment (AlanCoding) [inline awx/main/management/commands/run_callback_receiver.py:190]: @chrismeyersfsu was very interested in this problem. I expected that the job will be marked as finished before the final event comes in sometimes. They might alternate randomly. Solutions we discussed were scheduling a later task, but that creates all kinds of problems with reconciliation.

This is a simply retry 5 x 1 seconds. No matter what solution we have, some of them will be lost, and I don't want to spend a lot of time on the 0.001% of realistic cases. This approach simply ties up a callback worker for 5 seconds. I think that sounds fine.

Comment (ryanpetrello) [reply awx/main/management/commands/run_callback_receiver.py:190]: Yep, I saw situations where it went back and forth - sometimes the job _was_ finished, and sometimes it wasn't.  In cases where the events have all been processed but the job is _not_ `status=finished` yet, it's likely a _really small_ number of events (like the Hello, World! example).  In practice, I wouldn't expect real-world playbooks to hit this scenario very often.

> This approach simply ties up a callback worker for 5 seconds. I think that sounds fine.

Well, for _up to_ five seconds when a job finishes (in the scenario where the events are processed _faster_ than the celery worker can update the UnifiedJob status).  We're waiting up to five seconds for a different process to commit to the database.  Maybe that's too long?

Comment (AlanCoding) [reply awx/main/management/commands/run_callback_receiver.py:190]: If he gives this 👍 then I do too

File: awx/main/scheduler/task_manager.py [hunk 2]
```diff
@@ -630,4 +630,4 @@ def schedule(self):
 
                 # Operations whose queries rely on modifications made during the atomic scheduling session
                 for wfj in WorkflowJob.objects.filter(id__in=finished_wfjs):
-                    awx_tasks._send_notification_templates(wfj, 'succeeded' if wfj.status == 'successful' else 'failed')
+                    wfj.send_notification_templates('succeeded' if wfj.status == 'successful' else 'failed')
```
Comment (AlanCoding) [inline awx/main/scheduler/task_manager.py:633]: +1 this file looks good to me

**Review by wwitzel3 — APPROVED** [review]

**Comment by chrismeyersfsu** [comment]
To incorporate events processed count logic here we could use a transaction on post-commit hook to ensure that we get an accurate count of total events.

But by using event counts we would need to account for jobs where processed event count != expected event count. For that case we would need some sort of cleanup:
* periodic
* scheduled celery task per-job that can not fail (retry until success?)
* other

The idea here would need to be extended to workflows.

Brainstorming, might be better to have the task manager handle all of this work. Add another field on jobs that denotes "actually finished", or a property that effectively computes this. Then, in the scheduler (since it's periodic and more), have it emit the notifications and fire off the next wfj when necessary. We already do this for wfj notifications so we know the pattern.

Might need a notifications_sent field on the jobs too to denote notifications were processed.

**Comment by ryanpetrello** [comment]
@chrismeyersfsu is this an issue for workflows, though?  They don't really _have_ events to wait on, do they?  I don't think workflows should be susceptible to any sort of race, because they just report the name and _status_ of all of their nodes, which should be finalized but the time the task scheduler actually runs https://github.com/ansible/awx/pull/1652/files#diff-8f5e2fe91e83c971fc84cb35476362dbR633.  

In contrast, *normal* job launches are racy because they report their `job_host_summaries`, which is composed via the still-in-flight job events.

From `WorkflowJob`:

```python
def notification_data(self):
    result = super(WorkflowJob, self).notification_data()
    str_arr = ['Workflow job summary:', '']
    for node in self.workflow_job_nodes.all().select_related('job'):
        if node.job is None:
            node_job_description = 'no job.'
        else:
            node_job_description = ('job #{0}, "{1}", which finished with status {2}.'
                                    .format(node.job.id, node.job.name, node.job.status))
        str_arr.append("- node #{0} spawns {1}".format(node.id, node_job_description))
    result['body'] = '\n'.join(str_arr)
    return result
```

> Might need a notifications_sent field on the jobs too to denote notifications were processed.

One thing I like about my approach in this PR is that it delegates the processing to **one and only one** event that should always exist (the `EOF` event), so there's not any notion of "make sure we only process this once" like there could be if we implemented this in the task manager.

**Comment by AlanCoding** [comment]
My recommendation would be to get this in the current form right now, and then we could look at some more detailed error handling in generating the notification data, and generally be on the lookout for other problems.

**Pull request merged**